src/lapack/backends/cusolver/cusolver_batch.cpp

```diff
@@ -79,251 +100,645 @@ void getri_batch(sycl::queue &queue, std::int64_t n, sycl::buffer<std::complex<d
+        TYPE **tau, std::int64_t group_count, std::int64_t *group_sizes, TYPE *scratchpad,       \
+        std::int64_t scratchpad_size, const std::vector<sycl::event> &dependencies) {            \
+        return geqrf_batch(#CUSOLVER_ROUTINE, CUSOLVER_ROUTINE, queue, m, n, a, lda, tau,        \
+                           group_count, group_sizes, scratchpad, scratchpad_size, dependencies); \
+    }
+
+GEQRF_BATCH_LAUNCHER_USM(float, cusolverDnSgeqrf)
+GEQRF_BATCH_LAUNCHER_USM(double, cusolverDnDgeqrf)
+GEQRF_BATCH_LAUNCHER_USM(std::complex<float>, cusolverDnCgeqrf)
+GEQRF_BATCH_LAUNCHER_USM(std::complex<double>, cusolverDnZgeqrf)
+
+#undef GEQRF_BATCH_LAUNCHER_USM
+
+template <typename Func, typename T>
+inline sycl::event getrf_batch(const char *func_name, Func func, sycl::queue &queue, std::int64_t m,
+                               std::int64_t n, T *a, std::int64_t lda, std::int64_t stride_a,
+                               std::int64_t *ipiv, std::int64_t stride_ipiv,
+                               std::int64_t batch_size, T *scratchpad, std::int64_t scratchpad_size,
+                               const std::vector<sycl::event> &dependencies) {
+    using cuDataType = typename CudaEquivalentType<T>::Type;
+
+    overflow_check(m, n, lda, stride_a, stride_ipiv, batch_size, scratchpad_size);
+
+    // cuSolver legacy api does not accept 64-bit ints.
+    // To get around the limitation.
+    // Allocate memory with 32-bit ints then copy over results
+    std::uint64_t ipiv_size = stride_ipiv * batch_size;
+    int *ipiv32 = (int *)malloc_device(sizeof(int) * ipiv_size, queue);
+    int *devInfo = (int *)malloc_device(sizeof(int) * batch_size, queue);
+
+    auto done = queue.submit([&](sycl::handler &cgh) {
+        int64_t num_events = dependencies.size();
+        for (int64_t i = 0; i < num_events; i++) {
+            cgh.depends_on(dependencies[i]);
+        }
+        onemkl_cusolver_host_task(cgh, queue, [=](CusolverScopedContextHandler &sc) {
+            auto handle = sc.get_handle(queue);
+            auto a_ = reinterpret_cast<cuDataType *>(a);
+            auto devInfo_ = reinterpret_cast<int *>(devInfo);
+            auto scratchpad_ = reinterpret_cast<cuDataType *>(scratchpad);
+            auto ipiv_ = reinterpret_cast<int *>(ipiv32);
+            cusolverStatus_t err;
+
+            // Uses scratch so sync between each cuSolver call
+            for (int64_t i = 0; i < batch_size; ++i) {
+                CUSOLVER_ERROR_FUNC_T_SYNC(func_name, func, err, handle, m, n, a_ + stride_a * i,
+                                           lda, scratchpad_, ipiv_ + stride_ipiv * i, devInfo_ + i);
+            }
+        });
+    });
+
+    // Copy from 32-bit USM to 64-bit
+    sycl::event done_casting = queue.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(done);
+        cgh.parallel_for(sycl::range<1>{ ipiv_size }, [=](sycl::id<1> index) {
+            ipiv[index] = static_cast<std::int64_t>(ipiv32[index]);
```

Comment:
Original ipiv could be passed as *int32_t, and then resulting 32-bit values could be expanded to 64-bit (in-place).

Reply:
Could you explain further how the 32-bit values can be expanded to 64-bit in place? Are you suggesting use the 64-bit memory as if it were 32-bit then sequentially shift each value to the correct place?

Reply:
Yes, this is the idea. It would reduce amount of code (especially in array-of-pointers case), and runtime overhead.

Reply:
Okay, this would be an interested and likely more efficient approach. 

I think however it should be separated to this PR as the non-batched operations use a kernel for casting between types, same as here, so this change would be out of scope of this batch PR. Additionally, the change would be done for performance reasons so it would be good to measure if there is any performance benefit to the expanding method before implementing.

My thought is that it may make sense to create an issue to track this change so a future PR can be made once performance measurements have confirmed this approach is better and the PR can address both batched and non-batched implementations. Would this approach work for you?

Reply:
Batch case is different from regular by the size of problems: regular API is intended for larger cases, and this pre/post call overheads amortized by long kernel computations time. Batch case works on smaller problems, and run-time is also pretty short (due to high parallelism allowing to handle lots of small problems simultaneously). So any additional overhead is significant. Do you have any measurements of corresponding CUDA API on NV HW to compare it with the performance done through these SYCL interfaces?

Reply:
> Would this approach work for you?

Would work fine.

Reply:
> Do you have any measurements of corresponding CUDA API on NV HW to compare it with the performance done through these SYCL interfaces?

I currently do not have any measurements to compare the performance of this backend to native cuSolver

Reply:
Issue to track future change: https://github.com/oneapi-src/oneMKL/issues/230